classes/user_exception_selector.php

```diff
@@ -0,0 +1,193 @@
+     *                    If no users match the provided `$search` string, an empty array is returned instead (without nesting).
+     * @throws coding_exception
+     * @throws dml_exception
+     */
+    public function find_users($search): array {
+        global $DB;
+        // By default, `wherecondition` retrieves all users except the deleted, not confirmed and guest.
+        [$wherecondition, $params] = $this->search_sql($search, 'u');
+        $params['courseid'] = $this->courseid;
+        // Different queries depending on whether we want to search among users with or without existing exceptions in the course.
+        if ($this->skipauth) {
+            $stringkey = 'users_with_exception';
+            $ejoin = 'JOIN';
+        } else {
+            $stringkey = 'users_without_exception';
+            $ejoin = 'LEFT JOIN';
+            $wherecondition .= ' AND e.id IS NULL';
+        }
+        // We write out the query including the `SELECT` clause once here to enable IDE insights via connected DB table definitions.
+        $countsql = "SELECT COUNT(1)
+                       FROM {user} u
+                       JOIN {user_enrolments} AS ue ON ue.userid = u.id
+                       JOIN {enrol} AS en ON ue.enrolid = en.id
+                       JOIN {course} AS course ON en.courseid = course.id
+                     $ejoin {availability_shibboleth2fa_e} e ON (e.userid = u.id AND e.courseid = course.id AND e.skipauth = 1)
+                      WHERE $wherecondition
+                            AND course.id = :courseid";
+        if (!$this->is_validating()) {
+            $userscount = $DB->count_records_sql($countsql, $params);
+            if ($userscount > $this->maxusersperpage) {
+                return $this->too_many_results($search, $userscount);
+            }
+        }
+        $sql = preg_replace(
+            pattern: '/^SELECT COUNT\(1\)/',
+            replacement: "SELECT {$this->required_fields_sql('u')}",
+            subject: $countsql,
+            limit: 1,
+        );
+        [$sort, $sortparams] = users_order_by_sql('u', $search, $this->accesscontext);
+        $users = $DB->get_records_sql("$sql ORDER BY $sort", $params + $sortparams);
+        if (empty($users)) {
+            return [];
+        }
+        if ($search) {
+            $groupname = get_string("{$stringkey}_matching", 'availability_shibboleth2fa', $search);
+        } else {
+            $groupname = get_string($stringkey, 'availability_shibboleth2fa');
+        }
+        return [$groupname => $users];
+    }
+
+    /**
+     * Returns the options needed to recreate the given {@see user_exception_selector}.
+     *
+     * @return array Options to be passed to the class' constructor.
+     */
+    protected function get_options(): array {
+        return parent::get_options() + [
+            'courseid' => $this->courseid,
+            'skipauth' => $this->skipauth,
+        ];
```

Comment:
Don't we need the `'file'` field?

Reply:
No, see my comment in the commit message and the PR text above. It would be used [here](https://github.com/moodle/moodle/blob/ecfcf00d92df2e874da9d8cb31750403ada7789a/user/selector/search.php#L56-L60), but since Moodle uses its `PSR-0`-esque autoloading and our class is picked up by the autoloader (unlike its abstract parent class, which sits in a `lib.php` file), the class name provided by the parent method is sufficient.